src/Ledger/Epoch/Properties/GovDepsMatch.lagda

```diff
@@ -0,0 +1,184 @@
+open import Ledger.Ledger txs abs
+open import Ledger.Ledger.Properties txs abs
+open import Ledger.Prelude renaming (map to map'; mapˢ to map)
+open import Ledger.Ratify txs hiding (vote)
+open import Ledger.Utxo txs abs
+
+open import Axiom.Set.Properties th
+
+open import Data.List.Base using (filter)
+open import Data.List.Membership.Propositional.Properties using (∈-filter⁺; map-∈↔)
+open import Data.List.Subpermutations using (∈ˡ-map-filter)
+open import Data.Product.Function.NonDependent.Propositional using (_×-cong_)
+open import Data.Product.Properties using (×-≡,≡←≡)
+open import Data.Product.Properties.Ext using (×-⇔-swap)
+
+open import Relation.Unary using (Decidable)
+
+open Equivalence
+
+import Function.Related.Propositional as R
+import Relation.Binary.Reasoning.Setoid as SetoidReasoning
+
+module EPOCH-Body (eps : EpochState) where
+  open EpochState eps hiding (es) renaming (ls to epsLState; fut to epsRState) public
+  open RatifyState renaming (es to ensRState) public
+  open LState epsLState public
+  open GovActionState public
+  open UTxOState public
+
+  ens      = record (epsRState .ensRState) { withdrawals = ∅ }
+  tmpGovSt = filter (λ x → ¿ proj₁ x ∉ map proj₁ (epsRState .removed) ¿) govSt
+  orphans  = fromList $ getOrphans ens tmpGovSt
+  removed' = (epsRState .removed) ∪ orphans
+  removedGovActions = flip concatMapˢ removed' λ (gaid , gaSt) →
+    map (returnAddr gaSt ,_) ((deposits utxoSt ∣ ❴ GovActionDeposit gaid ❵) ˢ)
+
+module EPOCH-PROPS {eps : EpochState} where
+  open EPOCH-Body eps
+  open EpochState
+\end{code}
+
+\begin{lemma}[%
+  \LedgerMod{\EpochPropGov.lagda}{\AgdaModule{\EpochPropGov{}}}:
+  \AgdaFunction{govDepsMatch} is invariant of \EPOCH{} rule%
+  ]
+  \label{lem:EpochGovDepsMatch}
+
+  \begin{itemize}
+    \item \textit{Informally}.
+      Let \ab{eps}, \ab{eps'}~:~\EpochState{} be two epoch states and let
+      \ab{e}~:~\Epoch{} be an epoch.  Recall, \ab{eps}~\AgdaField{.ls} denotes the
+      ledger state of \ab{eps}.  If \ab{eps}~\AgdaDatatype{⇀⦇}~\ab{e}~\AgdaDatatype{,EPOCH⦈}~\ab{eps'}, then
+      (under a certain special condition) \govDepsMatch{}~(\ab{eps}~\AgdaField{.ls})
+      implies \govDepsMatch{}~(\ab{eps'}~\AgdaField{.ls}). 
+    \\[4pt]
+    To understand the special condition under which the property holds,
+    let \AgdaFunction{removed'} be the union of
+    \begin{itemize}
+      \item the governance actions in the \AgdaField{removed} field of the ratify
+    state of \ab{eps}, and
+      \item the orphaned governance actions in the \GovState{} of \ab{eps}.
+    \end{itemize}
+    Let $\mathcal{G}$ be the set
+    $\{\mbox{\GovActionDeposit{}~\ab{id}} : \mbox{\ab{id}} ∈ \mbox{proj}₁~\mbox{\AgdaFunction{removed'}}\}$.
+    The special hypothesis under which \govDepsMatch{} is an invariant of the
+    \EPOCH{} rule is the following: $\mathcal{G}$ is a subset of the set of deposits
+    of (the governance state of) \ab{eps}.
+    \item \textit{Formally}.
```

Comment:
Maybe get rid of the projections?

Reply:
Actually, the projections on the right can be replaced by using `dom`, so I'll do that, but I'd like to leave attempts to remove the projection on the left for another PR.